File: game-server/src/main/java/com/ggumtle/ggumtle/server/pipeline/PacketEncoder.java
package com.ggumtle.ggumtle.server.pipeline;

import com.ggumtle.ggumtle.server.packet.Packet;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class PacketEncoder extends MessageToByteEncoder<Packet> {

    @Override
    protected void encode(ChannelHandlerContext ctx, Packet packet, ByteBuf out) throws Exception {
        // 헤더를 바이트로 변환
        byte[] headerBytes = packet.header().toBytes();
        out.writeBytes(headerBytes);

        // 데이터가 있으면 추가
        if (packet.hasData()) {
            out.writeBytes(packet.data());
        }

        log.debug("패킷 인코딩 완료 - Type: {}, Length: {}", packet.header().packetType(), packet.getDataLength());
    }
}

File: game-load-test/src/main/kotlin/com/ggumtle/loadtest/protocol/codec/PacketEncoder.kt
package com.ggumtle.loadtest.protocol.codec

import com.ggumtle.loadtest.protocol.Packet
import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToByteEncoder
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

/**
 * Encodes Packet to bytes for network transmission
 */
class PacketEncoder : MessageToByteEncoder<Packet>() {

    override fun encode(ctx: ChannelHandlerContext, packet: Packet, out: ByteBuf) {
        // Write header (14 bytes)
        packet.header.writeTo(out)

        // Write data if present
        packet.data?.let { out.writeBytes(it) }

        logger.trace {
            "Encoded packet: type=${packet.header.packetType}, len=${packet.header.dataLength}"
        }
    }
}
